[GStreamer][Nexus][Broadcom] Added some platform quirks

calvaris opened the pull request:
#### 1d2768afac15d7a603e14fd43eeb30e20c64cf1d
<pre>
[GStreamer][Nexus][Broadcom] Added some platform quirks
<a href="https://bugs.webkit.org/show_bug.cgi?id=251789">https://bugs.webkit.org/show_bug.cgi?id=251789</a>

Reviewed by Philippe Normand.

Ensure we don&apos;t plug the wrong GStreamer element in AudioFileReader.
Consider nexus parsers as decoders because they decode but also parse and they are declared that way. Besides, nexus are
always decodes on hardware.
Add disclaimer comment for audio sink autoplug.
Add an environment variable to specify which video sink to load.
Add several options to WPE to enable different platform quirks on build.

Canonical link: <a href="https://commits.webkit.org/260064@main">https://commits.webkit.org/260064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e6864cd1c38c91bd1ecd37bd367f5369f99a080

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106919 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15955 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39716 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116099 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7153 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99105 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112684 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96221 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95089 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27865 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/96426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9103 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/95881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7047 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9680 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/95881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15270 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48775 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104654 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6944 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11210 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25924 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -36,6 +36,14 @@
+    GST_AUTOPLUG_SELECT_TRY,
+    GST_AUTOPLUG_SELECT_EXPOSE,
+    GST_AUTOPLUG_SELECT_SKIP
+} GstAutoplugSelectResult;
```
What is the point of redefining this?

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1211,11 +1211,24 @@ void MediaPlayerPrivateGStreamer::loadingFailed(MediaPlayer::NetworkState networ
+    GstElement* audioSink;
+
+#if PLATFORM(BCM_NEXUS)
+    audioSink = makeGStreamerElement("brcmaudiosink", nullptr);
```
This should not be needed. `autoaudiosink` should be able to create this... thing if it's properly ranked and registered as an audio sink.

calvaris replied:
I think they want to ensure you're using that sink only and if it is not present, you crash.

philn replied:
Can you confirm this please?

calvaris replied:
Do you think it would make sense to do the same scheme as for the video one? An env var?

philn replied:
Having your audio sink selected by autoaudiosink is not rocket science.
For video we can't really use autovideosink, that's why i suggested an env var for that.

calvaris replied:
It looks like we can go with auto plugged.

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3735,6 +3748,11 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSinkGL()
 
+#if PLATFORM(QCOM_DB)
+    sink = makeGStreamerElement("db410csink", "optimized vsink");
+#el#if PLATFORM(BCM_NEXUS)
```
typo

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3735,6 +3748,11 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSinkGL()
     }
 
+#if PLATFORM(QCOM_DB)
+    sink = makeGStreamerElement("db410csink", "optimized vsink");
```
Could this be done at runtime with an env var like `WEBKIT_GST_CUSTOM_VIDEO_SINK=mycooloptimizedsink`?

calvaris replied:
Why would you want that? I guess it's their preference to have it this way.

philn replied:
Reduced technical debt for us.

philn commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -133,6 +152,9 @@ AudioFileReader::~AudioFileReader()
     if (m_decodebin) {
         g_signal_handlers_disconnect_matched(m_decodebin.get(), G_SIGNAL_MATCH_DATA, 0, 0, nullptr, nullptr, this);
+#if PLATFORM(BCM_NEXUS)
+        g_signal_handlers_disconnect_matched(m_decodebin.get(), G_SIGNAL_MATCH_FUNC, 0, 0, nullptr, gpointer(decodebinAutoplugSelectCallback), nullptr);
```
`gpointer(foo)` expands to `void*(foo)` iiuc, is this a valid cast?

calvaris replied:
it seems to

calvaris replied:
`g_signal_handlers_disconnect_matched` takes a `gpointer` as a parameter, so it's almost ok

philn commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -383,6 +405,9 @@ void AudioFileReader::decodeAudioForBusCreation()
 
     m_decodebin = makeGStreamerElement("decodebin", "decodebin");
+#if PLATFORM(BCM_NEXUS)
+    g_signal_connect_swapped(m_decodebin.get(), "autoplug-select", G_CALLBACK(decodebinAutoplugSelectCallback), nullptr);
```
Why is this `_swapped`? The callback args don't look swapped to me.

calvaris replied:
True, and you made me wonder why it was working; because they are not used...

philn replied:
The `factory` argument is used.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -616,8 +616,11 @@ GstBuffer* gstBufferNewWrappedFast(void* data, size_t length)
-    if (!element)
+    if (!element) {
         WTFLogAlways("GStreamer element %s not found. Please install it", factoryName);
+        GST_WARNING("GStreamer element %s not found. Please install it", factoryName);
```
Why this new warning? `gst_element_factory_make()` already emits one when its return value is NULL :)

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -616,8 +616,11 @@ GstBuffer* gstBufferNewWrappedFast(void* data, size_t length)
+    if (!element) {
         WTFLogAlways("GStreamer element %s not found. Please install it", factoryName);
+        GST_WARNING("GStreamer element %s not found. Please install it", factoryName);
+        ASSERT_NOT_REACHED();
```
Also I thought there was agreement about this function not asserting, because it's call sites are supposed to handle NULL return values. This is for optional elements, basically.

calvaris committed:
[GStreamer][Nexus][Broadcom] Added some platform quirks
https://bugs.webkit.org/show_bug.cgi?id=251789

Reviewed by Philippe Normand.

Ensure we don't plug the wrong GStreamer element in AudioFileReader.
Consider nexus parsers as decoders because they decode but also parse and they are declared that way. Besides, nexus are
always decodes on hardware.
Add disclaimer comment for audio sink autoplug.
Add an environment variable to specify which video sink to load.
Add several options to WPE to enable different platform quirks on build.

Canonical link: https://commits.webkit.org/260064@main
